ggml/src/ggml-cpu/llamafile/sgemm.cpp

```diff
@@ -1049,6 +1049,577 @@ class tinyBLAS_Q0_AVX {
+                comparray[i] = 0;
+                int ca = 0;
+                const int8_t *at = aoffset->qs;
+                for (int j = 0; j < 32; j++)
+                    ca += (int)*at++;
+                comparray[i] = ca;
+                aoffset += lda;
+            }
+            COMPUTE(&acc_0, 0, 0);
+            COMPUTE(&acc_1, 4, 4);
+            COMPUTE(&acc_2, 0, 8);
+            COMPUTE(&acc_3, 4, 12);
+        }
+        SAVE_RES(ii, jj, 4, 4, 0);
+        SAVE_RES(ii+4, jj, 4, 4, 4);
+        SAVE_RES(ii, jj+4, 4, 4, 8);
+        SAVE_RES(ii+4, jj+4, 4, 4, 12);
+        free(comparray);
+    }
+
+    void gemm_small(int64_t m0, int64_t m, int64_t n0, int64_t n, int RM, int RN) {
+        int64_t ytiles = (m - m0) / RM;
+        int64_t xtiles = (n - n0) / RN;
+        int64_t tiles = xtiles * ytiles;
+        int64_t duty = (tiles + nth - 1) / nth;
+        int64_t start = duty * ith;
+        int64_t end = start + duty;
+        vec_t vec_A[8], vec_B[8] = {0};
+        vector signed int vec_C[4];
+        acc_t acc_0;
+
+        if (end > tiles)
+            end = tiles;
+        for (int64_t job = start; job < end; ++job) {
+            int64_t ii = m0 + job / xtiles * RM;
+            int64_t jj = n0 + job % xtiles * RN;
+            //int *comparray = (int *)malloc(RM * sizeof(int), 4096);
+            int *comparray = (int *)malloc(RM * sizeof(int));
+            vector float res[4] = {0};
+            vector float fin_res[4] = {0};
+            vector float vs[4] = {0};
+            vector float CA[4] = {0};
+            __builtin_prefetch((A+(ii*lda)+0)->qs, 0, 1); // prefetch first value
+            __builtin_prefetch((B+(jj*ldb)+0)->qs, 0, 1); // prefetch first value
+            for (int l = 0; l < k; l++) {
+                __builtin_prefetch((A+(ii*lda)+(l+1))->qs, 0, 1); // prefetch one loop ahead
+                __builtin_prefetch((B+(jj*ldb)+(l+1))->qs, 0, 1); // prefetch one loop ahead
+                __builtin_mma_xxsetaccz(&acc_0);
+                packNormal<int8_t, vector signed char>((A+(ii*lda)+l), lda, RM, 8, (int8_t*)vec_A, false);
+                packNormal<uint8_t, vector unsigned char>((B+(jj*ldb)+l), ldb, RN, 8, (uint8_t*)vec_B, true);
+                for(int x = 0; x < 8; x+=4) {
+                    __builtin_mma_xvi8ger4pp(&acc_0, vec_A[x], vec_B[x]);
+                    __builtin_mma_xvi8ger4pp(&acc_0, vec_A[x+1], vec_B[x+1]);
+                    __builtin_mma_xvi8ger4pp(&acc_0, vec_A[x+2], vec_B[x+2]);
+                    __builtin_mma_xvi8ger4pp(&acc_0, vec_A[x+3], vec_B[x+3]);
+                }
+                for (int I = 0; I<RM; I++) {
+                    for (int J = 0; J<RN; J++) {
+                        *((float*)&vs[I]+J) = (unhalf((A+((ii+I)*lda)+l)->d) * unhalf((B+((jj+J)*ldb)+l)->d));
+                    }
+                }
+                __builtin_mma_disassemble_acc(vec_C, &acc_0);
+                TA *aoffset = const_cast<TA*>(A+(ii*lda)+l);
```

Comment:
@Djip007 
When I try to cast it normally like this:
 TA *aoffset = (TA*)(A+(ii*lda)+l);
 I get this warning:
 warning: cast from type ‘const block_q8_0*’ to type ‘block_q8_0*’ casts away qualifiers [-Wcast-qual].
 
 Is there any other way to avoid this warning?

Reply:
I was thinking more about
```c++
const TA* aoffset = A+(ii*lda)+l;
```
ie: no need to remove the const. (and not a good idea for compiler optimisation)

```c++
auto aoffset = A + ii*lda + l;
```

Reply:
Done